core/src/main/java/org/apache/iceberg/schema/UnionByNameVisitor.java

```diff
@@ -179,20 +179,15 @@ private void updateColumn(Types.NestedField field, Types.NestedField existingFie
     }
   }
 
-  private boolean needsTypeUpdate(Type newType, Type existingType) {
-    if (newType.isPrimitiveType()) {
-      if (newType.equals(existingType)) {
-        return false;
-      }
-      if (existingType.typeId() == Type.TypeID.LONG && newType.typeId() == Type.TypeID.INTEGER) {
-        return false;
-      }
-      if (existingType.typeId() == Type.TypeID.DOUBLE && newType.typeId() == Type.TypeID.FLOAT) {
-        return false;
-      }
-      return true;
+  private boolean compatibleType(Type newType, Type existingType) {
+    switch (existingType.typeId()) {
+      case LONG:
+        return newType.typeId() == Type.TypeID.INTEGER || newType.typeId() == Type.TypeID.LONG;
```

Comment:
i'm even wondering here if we should have a intrinsic function for the type "isWiderThan(type)" or something ... but that may be overkill for now

Reply:
can we use [isPromotionAllowed ](https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/types/TypeUtil.java#L398)

Reply:
Ah great idea!

Reply:
Indeed excellent idea! 👍